apache/yunikorn-core
# [YUNIKORN-1467] [WIP] Implement priority-based victim selection algorithm

- Update to latest SI
- Implement getGuaranteedHeadRoom()
- Rename tryPreemption() -> tryRequiredNodePreemption()
- Update ask with both allowPreemptSelf / allowPreemptOther policies
- Remove preemption from partition conf
- Track last attempted preemption time on ask
- Walk preemption fence subtree to find potential victims
- Filter out allocations which don't help meet shortfall
- Filter out allocations which would reduce queue below guaranteed amounts
- Filter out nodes which have no allocations or cannot satisfy request
- Sort nodes by number of tasks which can be preempted
- Call shim to check predicates
- Limit number of preemption attempts to 10 requests per queue
- Revert YUNIKORN-1465 as we no longer need per-app preemption priorities
- Revert YUNIKORN-1466 as we no longer need per-queue preemption priorities

TODO:
- Unit testing (minimal unit tests present, not yet complete)
- E2E testing in shim (not started)
- Queue doesn't yet track resources for in-flight preemptions. This is necessary to avoid race conditions as we might preempt too much in a queue.
- Currently, preemption will only be triggered if there are no nodes capable of running a task. It will NOT be triggered in the case where only queue limits apply, as the initial check that a task is within limits prior to scheduling can't detect things like a sibling queue that is over guaranteed limits eating up a parent's max resources. Preemption should be allowed in this case, but the current node-centric algorithm doesn't handle this. We need some more design work on this scenario.

Issues:
- It appears that we get a reservation for a node even though that node is not a fit, resulting in the node being excluded from consideration due to the node iterator skipping reserved nodes.
- Configuration currently disallows parent guaranteed limits to be below the sum of child queue guarantees. Is this something we want to keep?

### What type of PR is it?

### Todos

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1467

### How should this be tested?
TODO 

### Screenshots (if appropriate)

### Questions:
